_posts/2021-11-07-canary.md

```diff
@@ -0,0 +1,123 @@
+If your Kafka cluster is exposed to external users you would like to know if they are able to use the cluster itself; the canary can be used to simulate activity for identifying problems from a user perspective even when the cluster seems to be operating correctly.
+Another use case could be when you are handling a fleet of Kafka clusters that you want to update to a newer version doing a canary rolling; just updating a few of them, checking that everything is fine and only then continuing with the rest of the fleet.
+
+These are the main problems that the canary tries to help with.
+Of course, it is not a replacement for any kind of monitoring solution that can be used to show metrics coming from the Kafka cluster, as provided by the corresponding Strimzi examples for setting up Prometheus and Grafana together with useful dashboards.
+
+Like the miners using the canaries, as part of the [sentinel species](https://en.wikipedia.org/wiki/Sentinel_species), in the coal mines for early warning about toxic gases in the 20th century, the Strimzi canary can be used to raise alerts when the Kafka cluster is starting to behave badly.
+
+## What the canary provides?
+
+The canary tool is made by several internal services, used for different purposes:
+
+* A `topic` service that is in charge to create and handle the dedicated topic (named `__strimzi_canary` by default) used by the canary producer and consumer to exchange messages.
+* A `producer` service that is used to send messages to the canary topic with a payload containing an identification number and mainly a timestamp, used to evaluate the end-to-end latency on the consumer side.
+* A `consumer` service that is used to receive the messages sent by the producer through the canary topic.
+* A `connection` service that just checks the connection to the brokers periodically; it opens and then immediately close a connection to each broker just to validate that it's reachable.
+* A `status` service that aggregates some of the producer and consumer metrics exposing them on a dedicated HTTP endpoint in a JSON format.
+
+When the canary starts for the first time, the `topic` service creates the dedicated topic with a number of partitions equals to the number of brokers, so that for each broker there is one partition.
+If the Kafka cluster is scaled up, the service adds more partitions to the topic assigning them in order to have the new partitions on the new brokers; the objective is always having one partition per broker so that when the producer and consumer exchange messages through the topic they are "testing" somehow all the brokers in the cluster using all partitions.
+If the Kafka cluster is scaled down, because it's not possible to remove partitions from a topic, the canary will just avoid using excess partitions (maybe reusing them if the cluster is scaled up again).
```

Comment:
`maybe reusing them` so I'm wondering if it is doing it or not?

Reply:
Is your point about removing `maybe`, to be clear that it's going to reuse them when scale up again?

Reply:
Yes. With `maybe` it kind of feels like it may or may not happen.